File: Java/DesignPatterns/ObjectPool.java
// Object Pool Demo in Java

// Helper Vector3D Class:
public class Vector3D
{
    public float x;
    public float y;
    public float z;

    public void Clean()
    {
        x = 0.0f;
        y = 0.0f;
        z = 0.0f;
    }

    public void Set(float x, float y, float z)
    {
        this.x = x;
        this.y = y;
        this.z = z;
    }
    
}
// Base GameObject Class:
public class GameObject
{
    public Vector3D position;
    public Vector3D rotation;
    public Vector3D scale;

    public GameObject()
    {
        position = new Vector3D();
        rotation = new Vector3D();
        scale = new Vector3D();
    }

    public void Clean()
    {
        position.Clean();
        rotation.Clean();
        scale.Clean();
    }
}

// The GameObject Pool:
import java.util.LinkedList;

public class GameObjectPool
{
    // Active and reserve list for the pool:
    LinkedList<GameObject> activeList;
    LinkedList<GameObject> reserveList;

    // Tracking data:
    int numberActive;
    int numberReserved;

    public GameObjectPool(int reserve = 5)
    {
        this.activeList = new LinkedList<GameObject>();
        this.reserveList = new LinkedList<GameObject>();

        this.numberActive = 0;
        this.numberReserved = 0;

        InitializeReserve(reserve);
    }

    // Helper: Initializes the reserve with X number of elements:
    private void InitializeReserve(int reserve)
    {
        for(int i = 0; i < reserve; i++)
        {
            GameObject gameObject = new GameObject();
            this.reserveList.addFirst(gameObject);
            this.numberReserved++;
        }
    }

    // Get a reserved GameObject:
    public GameObject GetGameObject()
    {
        if(this.numberReserved == 0)
        {
            // We can grow, or deny the request:
            // We'll grow:
            GameObject tmp = new GameObject();
            this.reserveList.addFirst(tmp);
            this.numberReserved++;
        }

        GameObject gameObject = this.reserveList.pop();
        this.numberReserved--;

        this.activeList.addFirst(gameObject);
        this.numberActive++;

        // Clean off the GameObject!
        gameObject.Clean();

        return gameObject;
    }

    // Recycle a GameObject that is no longer needed:
    public void ReturnGameObject(GameObject gameObject)
    {
        if(this.activeList.remove(gameObject))
        {
            this.numberActive--;

            this.reserveList.addFirst(gameObject);
            this.numberReserved++;
        }
    }
}

File: TypeScript/DesignPatterns/ObjectPool.ts
// The Vector3D and GameObject classes that make up the object pool entities:
class Vector3D
{
    public x: number;
    public y: number;
    public z: number;

    constructor()
    {
        this.clear();
    }

    public clear()
    {
        this.x = 0.0;
        this.y = 0.0;
        this.z = 0.0;
    }
}

class GameObject
{
    public position: Vector3D;
    public rotation: Vector3D;
    public scale: Vector3D;

    constructor()
    {
        this.position = new Vector3D();
        this.rotation = new Vector3D();
        this.scale = new Vector3D();
    }

    public clear()
    {
        this.position.clear();
        this.rotation.clear();
        this.scale.clear();
    }
}

// The object pool:
// Note:
/*
    Try implementing your own datastructure
    for the reserve and active lists!
    This way, you can guarantee the runtime complexity
    and have full control over your pool's efficiency.
*/
class GameObjectPool
{
    private activeList: Array<GameObject>;
    private reserveList: Array<GameObject>;

    private numberActive: number;
    private numberReserved: number;

    constructor(reserve: number = 5)
    {
        this.activeList = new Array<GameObject>();
        this.reserveList = new Array<GameObject>();

        this.numberActive = 0;
        this.numberReserved = 0;

        this.initializeReserve(reserve);
    }

    private initializeReserve(reserve: number)
    {
        for(let i = 0; i < reserve; i++)
        {
            const gameObject = new GameObject();
            this.reserveList.push(gameObject);
        }
    }

    public getGameObject(): GameObject
    {
        if(this.numberReserved == 0)
        {
            this.reserveList.push(new GameObject());
            this.numberReserved++;
        }

        const gameObject = this.reserveList.pop();
        this.numberReserved--;

        this.activeList.push(gameObject);
        this.numberActive++;

        gameObject.clear();

        return gameObject;
    }

    public returnGameObject(gameObject: GameObject)
    {
        // Get the index of the gameObject in the active list:
        const index = this.activeList.indexOf(gameObject);
        if(index >= 0)
        {
            // Splice the list around the element to remove.
            // Splice can be an expensive operation, which is why
            // I would use a custom collection in a real scenario:
            this.activeList.splice(index, 1);
            this.numberActive--;

            // Add it to the reserve:
            this.reserveList.push(gameObject);
            this.numberReserved++;
        }
    }
}